### PR TITLE
New Source interface draft

### PR DESCRIPTION
**Pull request by yao-msft**
- [ ] Have you signed the [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/winget-pkgs)?
- [ ] Are you working against an Issue?

-----

### CODE REVIEW

**Comment by yao-msft** [comment]
Changes mainly are:
1. Move repositorysearch.h implementation to itts own RepositorySearch.cpp
2. Created a new Source class which is more object oriented, which hosts an ISource(which becomes an internal interface)
3. Creation of the Source object just merely reads the configuration settings and acquires an ISource interface, which should be very lightweight
4. Most of the OpenSource logic are moved to Source::Open(ISource::Open), to decouple the source opening and source creation, thus more flexibility in source interactions.

File: src/AppInstallerRepositoryCore/Public/winget/RepositorySource.h [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+        std::vector<std::string> UnsupportedQueryParameters;
+
+        // Required query parameters in get manifest request.
+        std::vector<std::string> RequiredQueryParameters;
+    };
+
+    // Represents a source which would be interacted from outside of repository lib.
+    struct Source
+    {
+        // Default constructor to get all sources available. Can be used in source list, etc.
```
Comment (JohnMcPMS) [inline src/AppInstallerRepositoryCore/Public/winget/RepositorySource.h]: I think that this should create a `Source` object that cannot actually be acted on to enable the style:
```
Source a;
if (this) { a = Source("one"); }
else { a = Source("two"); }
```
Without having to do something expensive just to create `a` in the first place. #Resolved

File: src/AppInstallerRepositoryCore/Public/winget/RepositorySource.h [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+        operator bool() const;
+
+        // Gets the source's identifier; a unique identifier independent of the name
+        // that will not change between a remove/add or between additional adds.
+        // Must be suitable for filesystem names unless the source is internal to winget,
+        // in which case the identifier should begin with a '*' character.
+        const std::string GetIdentifier() const;
+
+        // Get the source's configuration details from settings.
+        const SourceDetails GetDetails() const;
```
Comment (JohnMcPMS) [inline src/AppInstallerRepositoryCore/Public/winget/RepositorySource.h]: `const &`? #Resolved

File: src/AppInstallerRepositoryCore/Public/winget/RepositorySource.h [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+        // Bool operator to check if a source reference is successfully acquired.
+        // Theoretically, the constructor could just throw when CreateSource returns empty.
+        // To avoid putting try catch everywhere, we use bool operator here.
+        operator bool() const;
+
+        // Gets the source's identifier; a unique identifier independent of the name
+        // that will not change between a remove/add or between additional adds.
+        // Must be suitable for filesystem names unless the source is internal to winget,
+        // in which case the identifier should begin with a '*' character.
+        const std::string GetIdentifier() const;
```
Comment (JohnMcPMS) [inline src/AppInstallerRepositoryCore/Public/winget/RepositorySource.h]: Either no `const` or `const &`. #Resolved

File: src/AppInstallerRepositoryCore/Public/winget/RepositorySource.h [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+
+        // Get the source's configuration details from settings.
+        const SourceDetails GetDetails() const;
+
+        // Get the source's information.
+        const SourceInformation GetInformation() const;
```
Comment (JohnMcPMS) [inline src/AppInstallerRepositoryCore/Public/winget/RepositorySource.h]: `const &`? #Resolved

File: src/AppInstallerRepositoryCore/Public/winget/RepositorySource.h [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+
+        // Set custom header.
+        bool SetCustomHeader(std::string_view header);
```
Comment (JohnMcPMS) [inline src/AppInstallerRepositoryCore/Public/winget/RepositorySource.h]:
```suggestion
        bool SetCustomHeader(std::optional<std::string_view> header);
```

Or offer a function to clear it. #Resolved

File: src/AppInstallerRepositoryCore/Public/winget/RepositorySource.h [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+        /* Composite sources */
+
+        // Gets a value indicating whether this source is a composite of other sources,
+        // and thus the packages may come from disparate sources as well.
+        bool IsComposite() const;
+
+        // Gets the available sources if the source is composite.
+        std::vector<Source> GetAvailableSources();
+
+        /* Writable sources */
```
Comment (JohnMcPMS) [inline src/AppInstallerRepositoryCore/Public/winget/RepositorySource.h:221]: These writing functions should probably stay on a separate derived type.

File: src/AppInstallerRepositoryCore/Public/winget/RepositorySource.h [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+        bool Add(IProgressCallback& progress);
+
+        // Update Source. Source update command.
+        std::vector<SourceDetails> Update(IProgressCallback& progress);
+
+        // Remove source. Source remove command.
+        bool Remove(IProgressCallback& progress);
+
+        // Drop source. Source reset command.
+        void Drop();
```
Comment (yao-msft) [inline src/AppInstallerRepositoryCore/Public/winget/RepositorySource.h:244]: >void Drop() [](http://example.com/codeflow?start=8&length=11)

static getallrefs() #Resolved

File: src/AppInstallerRepositoryCore/Public/winget/RepositorySource.h (hunk 3 above)
Comment (yao-msft) [reply src/AppInstallerRepositoryCore/Public/winget/RepositorySource.h]: no const

**Pull request closed without merging**